runner/websocket.go

```diff
@@ -0,0 +1,103 @@
+package runner
+
+import (
+	"bytes"
+	"compress/zlib"
+	"context"
+	"io"
+	"net/url"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/pganalyze/collector/output/pganalyze_collector"
+	"github.com/pganalyze/collector/state"
+	"github.com/pganalyze/collector/util"
+	"google.golang.org/protobuf/proto"
+)
+
+func SetupWebsocketForAllServers(ctx context.Context, servers []*state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) {
+	for idx := range servers {
+		go func(server *state.Server) {
+			logger = logger.WithPrefixAndRememberErrors(server.Config.SectionName)
+			for {
+				if server.WebSocket == nil {
+					connect(ctx, server, globalCollectionOpts, logger)
+				}
+				time.Sleep(60 * time.Second) // Delay between reconnect attempts
+			}
+		}(servers[idx])
+	}
+}
+
+func connect(ctx context.Context, server *state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) {
+	connCtx, cancelConn := context.WithCancel(ctx)
+	url, _ := url.Parse(server.Config.APIBaseURL + "/v2/websocket")
+	if url.Scheme == "http" {
+		url.Scheme = "ws"
+	} else {
+		url.Scheme = "wss"
+	}
+	headers := make(map[string][]string)
+	headers["Pganalyze-Api-Key"] = []string{server.Config.APIKey}
+	headers["Pganalyze-System-Id"] = []string{server.Config.SystemID}
+	headers["Pganalyze-System-Type"] = []string{server.Config.SystemType}
+	headers["Pganalyze-System-Scope"] = []string{server.Config.SystemScope}
+	headers["Pganalyze-System-Id-Fallback"] = []string{server.Config.SystemIDFallback}
+	headers["Pganalyze-System-Type-Fallback"] = []string{server.Config.SystemTypeFallback}
+	headers["Pganalyze-System-Scope-Fallback"] = []string{server.Config.SystemScopeFallback}
+	headers["User-Agent"] = []string{util.CollectorNameAndVersion}
+	headers["Sec-WebSocket-Protocol"] = []string{"websocket"}
+	conn, response, err := websocket.DefaultDialer.DialContext(connCtx, url.String(), headers)
+	if err != nil {
+		cancelConn()
+		time.Sleep(10 * time.Second)
+		logger.PrintError("Error starting websocket: %s %v", err, response)
```

Comment:
I wonder if there's a better output to show when doing a test run?

Reply:
Yeah, I was thinking it'd specifically be useful for enterprise users who e.g. have a proxy configured that doesn't support websockets.

Reply:
Maybe you can reuse `MarkCollectionAspectError` like:

```
server.SelfTest.MarkCollectionAspectError(state.CollectionAspectApiConnection, "error starting websocket with API: %s", err)
```

This is currently set by `GetDefaultGrant` but I think it won't be called via websocket way.